### PR TITLE
Update filtering capabilities for projects and enrollments

### PR DESCRIPTION
**Pull request by gigxz**
* Support filtering Projects by search term (used on Organization page list of projects)
  * Get rid of `ProjectSearchInput` and `projectSearch` query and move the search into the resolver. We want to be able to do searches on related fields (like searching on Organization.projects) so IMO this is easier to maintain and work with.
* Support filtering Enrollments by project type (not currently used, but can be a filter on a client's All Enrollments table)
* Support filtering Enrollments by client name search term (used on Project page list of enrollments)

 
Screen Shot 2023-01-17 at 12 25 05 PM
Screen Shot 2023-01-17 at 12 17 48 PM

### CODE REVIEW

File: drivers/hmis/app/graphql/types/hmis_schema/has_enrollments.rb [hunk 1]
```diff
@@ -33,9 +35,11 @@ def resolve_enrollments(scope = object.enrollments, **args)
 
       private
 
-      def scoped_enrollments(scope, sort_order: :most_recent, include_in_progress: false)
+      def scoped_enrollments(scope, sort_order: :most_recent, include_in_progress: false, project_types: nil, client_search_term: nil)
         scope = scope.viewable_by(current_user)
         scope = scope.where.not(project_id: nil) unless include_in_progress
+        scope = scope.with_project_type(project_types) if project_types.present?
+        scope = scope.joins(:client).merge(Hmis::Hud::Client.matching_search_term(client_search_term)) if client_search_term.present?
```
Comment (gigxz) [inline drivers/hmis/app/graphql/types/hmis_schema/has_enrollments.rb:42]: Is it bad to join client here, since it's modifying the result thats returned?

File: drivers/hmis/app/graphql/types/hmis_schema/query_type.rb [hunk 2]
```diff
@@ -16,19 +16,19 @@ class HmisSchema::QueryType < Types::BaseObject
     include Types::HmisSchema::HasOrganizations
     include Types::HmisSchema::HasClients
 
-    projects_field :projects, type: Types::HmisSchema::Project.page_type, description: 'Get a list of projects'
+    projects_field :projects
 
     def projects(**args)
       resolve_projects(Hmis::Hud::Project.all, **args)
```
Comment (gigxz) [inline drivers/hmis/app/graphql/types/hmis_schema/query_type.rb:22]: This is now used in place of `project_search` query. The `HasProjects` resolver handles the search term and the viewability filter.

File: drivers/hmis/app/graphql/types/hmis_schema/has_enrollments.rb (hunk 1 above)
Comment (eanders) [reply drivers/hmis/app/graphql/types/hmis_schema/has_enrollments.rb:42]: I don't think it's a problem.  It does mean if you have a client search term, you'll be limited to enrollments with a client, but I'd think you'd always want that, and should always have a client.  If it bothers you, you could just join client on line 39 so it always happens.

File: drivers/hmis/app/graphql/types/hmis_schema/has_enrollments.rb [hunk 3]
```diff
@@ -12,12 +12,14 @@ module HasEnrollments
...
 
       class_methods do
-        def enrollments_field(name = :enrollments, description = nil, type: Types::HmisSchema::Enrollment.page_type, **override_options, &block)
+        def enrollments_field(name = :enrollments, description = nil, type: Types::HmisSchema::Enrollment.page_type, without_args: [], **override_options, &block)
           default_field_options = { type: type, null: false, description: description }
           field_options = default_field_options.merge(override_options)
           field(name, **field_options) do
             argument :sort_order, HmisSchema::EnrollmentSortOption, required: false
             argument :include_in_progress, GraphQL::Types::Boolean, required: false
+            argument :project_types, [Types::HmisSchema::Enums::ProjectType], required: false  unless without_args.include? :project_types
```
Comment (eanders) [inline drivers/hmis/app/graphql/types/hmis_schema/has_enrollments.rb:21]: I'd love to se "without_args" flipped to a positive, it's often easier to write the negative case to exclude something that otherwise is there, but it's really hard to follow `unless without include`.  Could it be something like:
```
def enrollments_field(name = :enrollments, description = nil, type: Types::HmisSchema::Enrollment.page_type, extra_args: [:project_types], **override_options, &block)
....
argument :project_types, [Types::HmisSchema::Enums::ProjectType], required: false if extra_args.include? :project_types
```

Comment (eanders) [reply drivers/hmis/app/graphql/types/hmis_schema/has_enrollments.rb:21]: or maybe better if it's just 
```
def enrollments_field(name = :enrollments, description = nil, type: Types::HmisSchema::Enrollment.page_type, included_args: [:sort_order, :include_in_progress, :project_types, :client_search_term], **override_options, &block)
```
and then all of the args have a check.
I'm not sold on either of those either.

**Review by eanders — APPROVED** [review]
Looks like some good cleanup!  Thanks.  A couple thoughts inline.

File: drivers/hmis/app/graphql/types/hmis_schema/has_enrollments.rb (hunk 3 above)
Comment (gigxz) [reply drivers/hmis/app/graphql/types/hmis_schema/has_enrollments.rb:21]: Good point, but I'm not sure yet if it's worth the change. I think the way it works often _is_ exclusionary in nature... like you add a new search term (say `client_search_term` for Enrollment) and there's ONE place it really doesn't make sense to use (resolving within the context of one client like `Client.enrollments(...)`) but everywhere else (`Project.enrollments(...)`, `Organization.enrollments(...)`, `enrollments(...)`) it DOES make sense. If that's a common pattern, I'd rather just exclude one  place than have to add it in everywhere, I think. Let's wait and see how it plays out to see whether it should be flipped from a negative to positive.

Comment (eanders) [reply drivers/hmis/app/graphql/types/hmis_schema/has_enrollments.rb:21]: Sounds good.

File: drivers/hmis/app/graphql/types/hmis_schema/has_enrollments.rb (hunk 1 above)
Comment (gigxz) [reply drivers/hmis/app/graphql/types/hmis_schema/has_enrollments.rb:42]: Ok. The GraphQL schema requires enrollment to have `client`, so if that were to happen it would already be throwing a 500.
